File: IP/Fg333saCheck/Fg333saCheck.py
from z3 import *
import numpy as np
#variables
pbuff = []
VER_pbuff = Array('VER_pbuff', IntSort(), IntSort())
rdLen = Int('rdLen')
bComSuc = Int('bComSuc')
bComSuc_next = Int('bComSuc_next')
cntLenRdS = Int('cntLenRdS')
cntLenRdS_next = Int('cntLenRdS_next')
totalLenRdS = Int('totalLenRdS')
totalLenRdS_next = Int('totalLenRdS_next')
cntLenRdO = Int('cntLenRdO')
cntLenRdO_next = Int('cntLenRdO_next')
totalLenRdO = Int('totalLenRdO')
totalLenRdO_next = Int('totalLenRdO_next')
frm = Int('frm')
frm_next = Int('frm_next')
cntCheckO = Int('cntCheckO')
cntCheckO_next = Int('cntCheckO_next')
cntCheckS = Int('cntCheckS')
cntCheckS_next = Int('cntCheckS_next')
cntHeadO = Int('cntHeadO')
cntHeadO_next = Int('cntHeadO_next')
cntHeadS = Int('cntHeadS')
cntHeadS_next = Int('cntHeadS_next')
cntUpdataO = Int('cntUpdataO')
cntUpdataO_next = Int('cntUpdataO_next')
cntUpdataS = Int('cntUpdataS')
cntUpdataS_next = Int('cntUpdataS_next')
totalCheckO = Int('totalCheckO')
totalCheckO_next = Int('totalCheckO_next')
totalCheckS = Int('totalCheckS')
totalCheckS_next = Int('totalCheckS_next')
totalHeadO = Int('totalHeadO')
totalHeadO_next = Int('totalHeadO_next')
totalHeadS = Int('totalHeadS')
totalHeadS_next = Int('totalHeadS_next')
totalUpdataO = Int('totalUpdataO')
totalUpdataO_next = Int('totalUpdataO_next')
totalUpdataS = Int('totalUpdataS')
totalUpdataS_next = Int('totalUpdataS_next')
def Array_Sum(a,k):
	temp = 0
	for i in range(k):
		temp += a[i]
	return temp
def Array_length(a):
	temp = np.array(a)
	return temp.size
def List2Array(lst, idx=()):
    if isinstance(lst[0], list):
        size = len(lst)
        z3_array = Array('array_' + '_'.join(map(str, idx)), IntSort(), ArraySort(IntSort(), IntSort()))
        constraints = []
        for i in range(size):
            sub_array, sub_constraints = List2Array(lst[i], idx + (i,))
            constraints.extend(sub_constraints)
            constraints.append(z3_array[i] == sub_array)
        return z3_array, constraints
    else:
        size = len(lst)
        z3_array = Array('array_' + '_'.join(map(str, idx)), IntSort(), IntSort())
        constraints = [z3_array[i] == lst[i] for i in range(size)]
        return z3_array, constraints
s = Solver()
#Contract

File: VeriHar-main/sources/Fg333saCheck.c
#include "VeriHar-main/include/Fg333saCheck.h"

unint08 CheckSumAdd08(const unint08 *buf, unint32 num) {
	unint32 sum = 0;
  for (unint32 i = 0; i < num; i++) {
		sum = sum + buf[i];
	}
	return (unint08)(sum%255);
}

/*
fg_3_3_3sa协议校验IP

功能描述：
        校验接收到的数据是否符合Fg333型光纤陀螺通信协议，校验成功返回TRUE32，否则返回FALSE32。
        1.校验数据长度是否正确：若长度不为19，则cntLenRd、totalLenRd
+1，返回校验失败；
        2.校验帧计数是否更新：若帧计数不更新，则cntUpdata、totalUpdata
+1，返回校验失败； 3.校验帧头是否正确：若帧头非0xEB90，则cntHead、totalHead
+1，返回校验失败；
        4.校验累加和是否正确：若累加和正确则返回校验成功，否则返回校验失败。
*/
void Fg333saCheckFun(void *pIp) {
  unint08 chksum;
  Fg333saCheck *p = (Fg333saCheck *)pIp;

  p->bComSuc = FALSE32;

  if (p->rdLen == 19) {
    p->cntLenRdS = 0;

    /* 帧计数是否更新 */
    if (p->pbuff[17] != (p->frm)) {
      p->cntUpdataS = 0;
      p->cntUpdataO = 0;
      p->frm = p->pbuff[17];

      /* 帧头为0xEB90 */
      if ((p->pbuff[0] == 0xEB) && (p->pbuff[1] == 0x90)) {
        p->cntHeadS = 0;
        p->cntHeadO = 0;
        /* 计算累加和 */
        chksum = CheckSumAdd08(&p->pbuff[0], 18);

        /* 累加和正确 */
        if (chksum == p->pbuff[18]) {
          p->cntCheckS = 0;
          p->cntCheckO = 0;
          p->bComSuc = TRUE32;
        } else {
          /* 累加和不正确 */
          p->cntCheckS++;
          p->totalCheckS++;
          p->cntCheckO++;
          p->totalCheckO++;
        }
      } else {
        /* 帧头不正确 */
        p->cntHeadS++;
        p->totalHeadS++;
        p->cntHeadO++;
        p->totalHeadO++;
      }
    } else {
      /* 帧计数不更新 */
      p->cntUpdataS++;
      p->totalUpdataS++;
      p->cntUpdataO++;
      p->totalUpdataO++;
    }
  } else {
    /* 数据长度不对 */
    p->cntLenRdS++;
    p->totalLenRdS++;
    p->cntLenRdO++;
    p->totalLenRdO++;
  }
  return;
}
